README.md
Zap Ruder
=========

Zap Ruder is a Zapper test program.

Background
----------
On November 22, 1963, Abraham Zapruder used a Bell & Howell 8mm
camera to shoot a 27-second silent short film of President John F.
Kennedy's limousine and managed to catch his assassination by a
sniper.  The developed film was used in the investigation of this
heinous murder.

The Zapper is a light gun for the NES made by Nintendo, commonly used
for the game *Duck Hunt*.  Just as Zapruder's camera recorded light
from the scene, the Zapper works by measuring light from the TV.
It contains a trigger switch and a photosensor that detects whether
or not the barrel is aimed at a bright area on the screen.

The photosensor is connected to a resonator that distinguishes light
from a CRT SDTV, which flickers at the 15.7 kHz horizontal scan rate,
from any other light source.  As soon as it detects the appropriate
rate, it activates a signal on the controller port.  Because this
time is proportional to how far down the barrel is pointed, one can
think of this signal as using [pulse-position modulation].  By using
timed code, the NES CPU can count how long it takes between the start
of a frame and when the photosensor begins to receive light and
estimate the barrel's position that way.

But very few NES homebrew games use the Zapper.  I suspect that this
might have three causes:

 1. Proliferation of LCD HDTVs, as the Zapper's photosensor works
    only with the CRT SDTVs popular during the NES's commercial era.
 2. Little published source code for how to make the most of
    the Zapper.
 3. Inaccurate emulation of the Zapper in popular emulators.

Cause 1 will only get worse, but I released Russian Roulette
as a tech demo of reading the trigger only.  Zap Ruder aims
to solve cause 2 and provide test cases for solving cause 3.

This demo is for NTSC NES and NTSC TV.  It should also run on RGB
modded NES.  It might even work on PAL famiclones such as Dendy for
what it's worth.  But expect noticeable mistracking on a PAL NES due
to a different CPU clock ratio.  The serial gun of *Vs. Duck Hunt*
is not yet supported.

[pulse-position modulation]: https://en.wikipedia.org/wiki/Pulse-position_modulation

The menu
--------
The menu demonstrates the common technique of flickering the targets
to determine which target the player is pointing at.  When the
photosensor is moved from a dark area to a light area, the menu
darkens the left column and the right column in sequence.  Then it
measures how far down the photosensor is using the "yonoff" kernel
(see below).

If this menu doesn't track well on your TV or your emulator, you
can use the Control Pad of the controller in port 1 instead.

Technical tests
---------------
All tests use a gun in port 2 and a standard controller in port 1
unless otherwise specified.  All tracking tests use the "yonoff"
kernel, which indicates how far down the screen the brightness
started and stopped being detected for the gun in port 2, unless
otherwise specified.

Basic tracking tests use a dithered background of two colors.
Hue is 0 (gray) or 1-12 (color), and brightness is 0 to 8.
Pattern tests use solid-colored objects; brightness can be set only
to odd values.  The colors can also be used to calibrate an NTSC TV's
tint knob: 4 should be magenta (R=B, minimal G), and 10 should be
green (G, minimal R=B).  Pulling the trigger twice rapidly will close
each tracking test.

### Y tracking

Most of the screen is filled with flat dither whose brightness and
hue can be adjusted.

### Two-gun Y tracking

Kernel: yon2p (dual gun)

Most of the screen is filled with flat dither whose brightness and
hue can be adjusted.  This test indicates the start of brightness for
guns in both controller ports; the end of brightness is not measured.

### X tracking

Kernel: xyon

Most of the screen is filled with flat dither whose brightness and
hue can be adjusted.  This test indicates both the horizontal and
vertical position of the start of brightness, though the horizontal
position is very noisy.

The intent of this experimental kernel was to estimate the horizontal
position in 18-pixel units, but noise caused the actual detection
onset to vary by six units.  Thus this kernel isn't very useful.

### Pattern tests

Most of the screen is filled with vertical or horizontal line
patterns whose brightness and hue can be adjusted, but brightness
can be set only to odd values (which correspond to flat colors).

1 of 8, 2 of 8, 3 of 8, 4 of 8
5 of 8, 6 of 8, 7 of 8, 8 of 8
1 of 4, 2 of 4, 3 of 4, 1 of 2

Color can be set to any hue and odd brightness; the rest is black.

### Ball tests

In the center of the screen is a circle whose size, brightness, and
hue can be adjusted, to show how small of a target the photosensor
can reliably detect.  As with the vertical and horizontal line
pattern tests, brightness can be set only to odd values.

### Trigger test

The Zapper contains a mechanism to release the trigger switch once
the trigger is pulled all the way, but one can keep the switch on by
holding the trigger halfway.  This is a simple test that counts
how long the trigger switch on the gun in port 2 is held before it
is released.  This is the only test that works on an HDTV.

Because the Zapper pulls the trigger down harder than the Control
Deck can pull it back up, the switch will usually appear active for
at least 5 frames.

Toys
----

### Axe

Point the gun at the screen to play notes.  Pitch is proportional
to height on a pentatonic scale.  Create rhythm by covering and
uncovering the barrel.  Axe will store your performance and echo
it back to you eight measures later.  Hold the trigger halfway to
change the timbre; press it quickly to accent a note.

To exit, shoot offscreen twice.

Games
-----

### ZapPing

Kernel: yon2p (dual gun)

You've played Pong, the primitive air hockey simulator that was the
first popular video game.  You may have played Odyssey, which was
Pong before Pong was cool.  You may have even played FlapPing for
Atari 2600.  But it is the 2010s, and there is time for ZapPing.
It manages to coax a control feel nearly as smooth as a Wii Remote
from 1985 technology.  But the basic rule of air hockey remains:
avoid missing ball for high score.

At the title screen, press A or pull the trigger to join.  Press the
button again to play against the NES, or press the button on the
other controller to play with two players.

The two player characters are named Podge, in gray, and Daffle, in
red.  These correspond to the two colors of Zapper that Nintendo
has sold, both before and after a change to United States toy safety
regulations.

Move your paddle up and down with the Zapper or by pressing Up or
Down on the Control Pad.  A players with a Zapper will have an unfair
advantage over a player with a controller, just as first-person
shooter players with a mouse have an advantage over players with an
Xbox controller.  Just remember not to point your Zapper off the
big green table because it will stop tracking until you point it at
the table again.  (The Arkanoid controller and Super NES Mouse are
not yet supported due to lack of time, and Power Glove controllers
are not supported yet due to lack of hardware with which to test.
Donations are welcome.)

Each player serves the ball with the A button or the trigger for
two balls before passing.  The game speeds up gradually.  When you
miss the ball, the other player gets a point, and the next serve will
lose some speed but quickly catch up.  The winner is the player with
at least 11 points and ahead by two, except at 20-20, 21 always wins.

To exit, press B on controller 1 at the title screen, or press Reset
if two guns are connected.

Contact
-------
Apart from filing an issue, the most reliable ways to contact the
maintainer about Zap Ruder are by replying to a [topic on NESdev BBS]
or by leaving comments on the project's [talk page].

[topic on NESdev BBS]: https://forums.nesdev.com/viewtopic.php?t=8108
[talk page]: https://pineight.com/mw/?title=Talk:Zap_Ruder

Legal
-----
Copyright 2012, 2019 Damian Yerrick

Copying and distribution of this file, with or without
modification, are permitted in any medium without royalty provided
the copyright notice and this notice are preserved in all source
code copies.  This file is offered as-is, without any warranty.



CHANGES.txt
0.03a (2019-04-15)
* Rebuild with Python 3 and recent ca65
* Move audio subroutine names into Pently namespace
* Move binaries from repository to release

0.03 (2012-05-14)
* Axe: a music toy
* music engine supports playing notes and performing a call back
* pads: disable DPCM-safe operation at compile time
* Can double-shoot offscreento leave pattern screens
* ZapPing: Press player 1 B button to leave this screen

0.02 (2011-09-30)
* pads: Zapper trigger presses the A Button, to simplify programming
* has a title screen and music
* ZapPing: a tennis game
* Build process produces an NSF

0.01 (2011-09-18)
* initial release
* kernel tests with solid color: Y on/off, 2-player Y, and X/Y
  (in X/Y, X is very noisy)
* pattern tests: horizontal lines, vertical lines, and ball patterns
* trigger hold time


src/axe.s
;
; Axe music toy for NES
; Copyright 2012 Damian Yerrick
;
; Copying and distribution of this file, with or without
; modification, are permitted in any medium without royalty provided
; the copyright notice and this notice are preserved in all source
; code copies.  This file is offered as-is, without any warranty.
;
.include "nes.inc"
.include "global.inc"
.export music_row_callback, music_dalsegno_callback, axe_callback_on

; overlay some of tennis's variables
.importzp paddle_yhi, paddle_ylo, ball_dxlo, ball_dxhi

AXE_SCROLL_SPEED = 16*256/7
SOFT_ACCENT_THRESHOLD = 10

.segment "BSS"
lastNotes: .res 128
rowNumber: .res 1
axe_callback_on: .res 1
axe_xoffset_lo = ball_dxlo
axe_xoffset_hi = ball_dxhi
noteAccent = paddle_ylo
explosion_y = paddle_yhi+1

scroll_colbuf = $0100
scroll_dsthi = $0118
scroll_dstlo = $0119

new_dot_shape = $011D
new_dot_dsthi = $011E
new_dot_dstlo = $011F

.segment "CODE"
.align 128
.proc zapkernel_dummy_ntsc
  lda #$00
loop:
  ; 96
  jsr waste_12
  jsr waste_12
  jsr waste_12
  jsr waste_12
  jsr waste_12
  jsr waste_12
  jsr waste_12
  jsr waste_12

  ; 6
  bit $00
  bit $00

  ; 11.667
  clc
  adc #$AA
  bcs :+
:
  dey
  bne loop
waste_12:
  rts
.endproc

.proc axe
  lda #VBLANK_NMI
  sta PPUCTRL
  sta doubleshoot_time
  sta axe_callback_on
  ldx #$00
  stx noteAccent
  stx held_time
  stx PPUMASK
  lda #$3F
  sta PPUADDR
  stx PPUADDR
copypal:
  lda axe_palette,x
  sta PPUDATA
  inx
  cpx #$20
  bcc copypal
  
  jsr axe_init_bg
  
  ldx #127
  lda #$FF
  sta rowNumber
clear_notes:
  sta lastNotes,x
  dex
  bpl clear_notes
  lda #3
  jsr pently_start_music
  lda #$FF
  sta new_dot_dsthi
  sta scroll_dsthi

loop:
  jsr prepare_vram_transfers
  
  lda nmis
:
  cmp nmis
  beq :-
  lda PPUSTATUS
  
  lda #VBLANK_NMI|VRAM_DOWN
  sta PPUCTRL
  lda scroll_dsthi
  bmi no_new_scroll
  sta PPUADDR
  lda scroll_dstlo
  sta PPUADDR
  ldx #23
scroll_blit_loop:
  lda scroll_colbuf,x
  sta PPUDATA
  dex
  bpl scroll_blit_loop

  lda #$FF
  sta scroll_dsthi
no_new_scroll:

  ; Draw new dot  
  lda new_dot_dsthi
  bmi no_new_dot
  sta PPUADDR
  lda new_dot_dstlo
  sta PPUADDR
  lda new_dot_shape
  sta PPUDATA
  lda #$FF
  sta new_dot_dsthi
no_new_dot:

  ; TO DO: draw new column of notes if needed

  ldx 2
  ldy #0
  sty OAMADDR
  lda #>OAM
  sta OAM_DMA  
  lda 3
  sec
  jsr ppu_screen_on
  lda #$FF
  sta scroll_dsthi
  
  jsr read_pads
  clc
  lda #<AXE_SCROLL_SPEED
  adc axe_xoffset_lo
  sta axe_xoffset_lo
  lda #>AXE_SCROLL_SPEED
  adc axe_xoffset_hi
  sta axe_xoffset_hi

  ; compute held_time for accent
  lda cur_keys+1
  ora cur_keys
  and #KEY_A
  sta cur_trigger
  beq notA
  lda #1
  sta noteAccent
  inc held_time
  lda held_time
  cmp #20
  bcc doneA
  lda #20
  bne doneA
notA:
  lda held_time
  beq doneA
  cmp #SOFT_ACCENT_THRESHOLD
  bcs noHardAccent
  lda #2
  sta noteAccent
noHardAccent:
  lda #0
doneA:
  sta held_time
  
  jsr pently_update

  ; read the zapper
  jsr s0wait
  ldy #191
  jsr zapkernel_yonoff_ntsc
  cpy #0
  beq :+
  jsr zapkernel_dummy_ntsc
:
  lda #VBLANK_NMI|BG_0000|OBJ_1000
  sta PPUCTRL
  lda #0
  sta PPUSCROLL
  sta PPUSCROLL

  lda 0
  ldx 1
  bne not_offscreen
  lda #$EF
not_offscreen:
  and #%11111000
  ora #%00000011
  sta paddle_yhi
  

  ; If player 2's gun has shot twice in quick succession
  ; and is pointed offscreen, cancel
  cpx #0
  bne not_triggered
  lda new_keys+1
  bpl not_triggered
  lda doubleshoot_time
  cmp #DOUBLESHOOT_FRAMES
  bcs triggered_once
  lda #KEY_B
  ora new_keys+0
  sta new_keys+0
triggered_once:
  lda #0
  sta doubleshoot_time
not_triggered:
  inc doubleshoot_time
  bne :+
  dec doubleshoot_time
:

  lda new_keys
  and #KEY_B
  bne done
  jmp loop
done:
  jsr pently_init
  lda #0
  sta axe_callback_on
  rts
.endproc

.proc music_row_callback
  bit axe_callback_on
  bmi :+
  rts
:

  lda #0
  sta axe_xoffset_lo
  sta axe_xoffset_hi
  lda #$FF
  sta explosion_y
  inc rowNumber
  lda rowNumber
  and #$7F
  sta rowNumber

; triangle: play 128 notes back
  sec
  ldy rowNumber
  lda lastNotes,y
  cmp #$FF
  beq no_minus128

  lsr a
  lsr a
  tax
  ldy #9  ; instrument for triangle echo
  lda pentatonic_notes,x
  clc
  adc #12
  ldx #8
  jsr pently_play_note
  
no_minus128:
  
  lda #190
  sec
  sbc paddle_yhi
  cmp #191
  bcs clear_echo
  lsr a
  lsr a
  lsr a
  tax
  
  ; prepare the note number to be drawn
  lda pentatonic_keycolors,x
  ora #$02
  sta new_dot_shape
  jsr get_dot_pos_x
  
  ; store the note number
  txa
  asl a
  asl a
  ora noteAccent
  ldy rowNumber
  sta lastNotes,y
  
  ; set the instrument (3-5: normal; 6-8: echoed)
  lda noteAccent
  clc
  adc #3
  tay
  lda pentatonic_notes,x
  ldx #4  ; square 2
  jsr pently_play_note

  dec noteAccent
  bpl no_clip_accent
  lda #0
  sta noteAccent
no_clip_accent:
  jmp no_minusthree
clear_echo:
  lda #$FF
  ldy rowNumber
  sta lastNotes,y

echo_minusthree:
  sec
  lda rowNumber
  sbc #3
  and #$7F
  tay
  lda lastNotes,y
  cmp #$FF
  beq no_minusthree

  lsr a
  lsr a
  tax
  lda lastNotes,y
  and #$03
  clc
  adc #6
  tay
  lda pentatonic_notes,x
  ldx #4
  jsr pently_play_note
  
no_minusthree:

  rts
.endproc

.proc music_dalsegno_callback
  lda #0
  sta rowNumber
  rts
.endproc

; BG GRAPHICS ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

.proc axe_init_bg
nts_left = 7
rows_left = 6

  ; First establish the stripes across the nametables.
  lda #$20
  sta PPUADDR
  lda #$00
  sta PPUADDR
  lda #2
  sta nts_left
ntloop:
  lda #0
  ldy #1
  jsr ppu_sta32y
  lda #'_'
  ldy #1
  jsr ppu_sta32y
  lda #23
  sta rows_left
rowloop:
  ldy rows_left
  lda pentatonic_keycolors,y
  ldy #1
  jsr ppu_sta32y
  dec rows_left
  bpl rowloop
  lda #0
  ldy #6
  jsr ppu_sta32y
  dec nts_left
  bpl ntloop

  ; now draw the bar lines for measures and beats
barline_type = rows_left
  lda #VBLANK_NMI|VRAM_DOWN
  sta PPUCTRL
  lda #$20
  sta nts_left
barline_ntloop:
  ldx #$47
barlineloop:

  ; beat dots in columns 7, 15, 23; measure dots in 31
  lda #$06
  cpx #$5F
  bcc barline_not_measure
  lda #$04
barline_not_measure:
  sta barline_type

  lda nts_left
  sta PPUADDR
  stx PPUADDR
  ldy #23
barcell_loop:
  lda pentatonic_keycolors,y
  ora barline_type
  sta PPUDATA
  dey
  bpl barcell_loop
  
  ; next barline
  txa
  clc
  adc #8
  tax
  cpx #$60
  bcc barlineloop

  ; next nametable  
  lda nts_left
  adc #4-1
  sta nts_left
  cmp #$28
  bcc barline_ntloop
  lda #VBLANK_NMI
  sta PPUCTRL
  rts
.endproc

TRAILING_ROWS = 6
VERTICAL_BAR_TILE = $09

.proc get_dot_pos_x
  lda #0
  sta new_dot_dstlo
  txa
  eor #$1F
  sec
  sbc #6
  ror a
  ror new_dot_dstlo
  lsr a
  ror new_dot_dstlo
  ror a
  ror new_dot_dstlo
  sta new_dot_dsthi
  lda rowNumber
  and #$0F
  asl a
  ora new_dot_dstlo
  sta new_dot_dstlo
  lda rowNumber
  and #$10
  lsr a
  lsr a
  ora new_dot_dsthi
  sta new_dot_dsthi
  txa
  asl a
  asl a
  asl a
  eor #$FF
  adc #202
  sta explosion_y
  rts
.endproc

.proc prepare_vram_transfers

  ; 1. Draw next column to be seen in 16 ticks
  lda rowNumber
  clc
  adc #16
  and #$7F
  tay
  lda lastNotes,y
  lsr a
  lsr a
  sta 0
  
  ldx #23
makecolloop:
  lda #0
  cpx 0
  bne :+
  lda #2
:
  ora pentatonic_keycolors,x
  sta scroll_colbuf,x
  dex
  bpl makecolloop

  ; 2. Compute destination address for next row 
  lda rowNumber
  and #$0F
  asl a
  ora #$40
  sta scroll_dstlo
  lda rowNumber
  and #$10
  lsr a
  lsr a
  eor #$24
  sta scroll_dsthi

  ; 3. Draw aiming cursor
  lda paddle_yhi
  clc
  adc #16
  tay
  ldx #8
  lda #0
  jsr draw_y_arrow_sprite
  
  ; 4. Draw explosion
  ldx oam_used
  lda explosion_y
  cmp #$EF
  bcs no_explosion

  sta OAM+0,x
  sta OAM+4,x
  sta OAM+8,x
  sta OAM+12,x
  sta OAM+16,x
  sta OAM+20,x
  lda #8
  sta OAM+1,x
  sta OAM+5,x
  sta OAM+9,x
  sta OAM+13,x
  sta OAM+17,x
  sta OAM+21,x
  lda #1
  sta OAM+2,x
  sta OAM+6,x
  sta OAM+10,x
  sta OAM+14,x
  sta OAM+18,x
  sta OAM+22,x
  lda #96
  clc
  adc axe_xoffset_hi
  sta OAM+3,x
  clc
  adc axe_xoffset_hi
  sta OAM+7,x
  clc
  adc axe_xoffset_hi
  sta OAM+11,x
  lda #96
  sec
  sbc axe_xoffset_hi
  sta OAM+15,x
  sec
  sbc axe_xoffset_hi
  sta OAM+19,x
  sec
  sbc axe_xoffset_hi
  sta OAM+23,x
  txa
  clc
  adc #24
  tax
no_explosion:

  ; 5. Draw timeline
  lda nmis
  and #$01
  beq timeline_flickering_top
  lda #$08
timeline_flickering_top:
  clc
  adc #15
  sta 2
  ldy #12
timeline_rowloop:
  lda 2
  sta OAM,x
  clc
  adc #16
  sta 2
  lda #VERTICAL_BAR_TILE
  sta OAM+1,x
  lda #%00000000
  sta OAM+2,x
  lda #16*TRAILING_ROWS
  sta OAM+3,x
  inx
  inx
  inx
  inx
  dey
  bne timeline_rowloop
  jsr ppu_clear_oam
  
  ; 6. Set up sprite 0
  lda #4
  sta oam_used
  lda #7
  sta OAM+0
  lda #4
  sta OAM+1
  lda #%00000000
  sta OAM+2
  lda #128
  sta OAM+3
  
  ; 7. Calculate scroll position
  lda #0
  sta 3
  lda rowNumber
  sec
  sbc #TRAILING_ROWS
  asl a
  rol 3
  asl a
  rol 3
  asl a
  rol 3
  asl a
  rol 3
  adc axe_xoffset_hi
  sta 2
  lda 3
  adc #0
  and #$01
  ora #VBLANK_NMI|BG_0000|OBJ_1000
  sta 3

  rts
.endproc

.segment "RODATA"
pentatonic_notes:
  .repeat 5, I
    .byt 4+12*I, 7+12*I, 9+12*I, 11+12*I, 14+12*I
  .endrepeat
pentatonic_keycolors:
  .repeat 5
    .byt $10, $11, $10, $10, $11
  .endrepeat
axe_palette:
  .byt $0F,$10,$21,$30, $0F,$10,$21,$30, $0F,$10,$21,$30, $0F,$10,$21,$30
  .byt $0F,$26,$26,$26, $0F,$16,$27,$38, $0F,$00,$00,$00, $0F,$00,$00,$00
